### ISSUE TITLE
"No Android SDK found" error when building Android demo app

### ISSUE THREAD

**Issue by daj** [tensorflow/tensorflow#7261]
I'm only following [the README instructions](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/android/README.md) to build the Android demo app.

I believe I have configured my ./WORKSPACE file correctly.  This seems to be a different error to the obvious one you get if you haven't update the WORKSPACE file...

### What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?

http://stackoverflow.com/q/41478820/112705
[No matching open or closed Issues in this Git project](https://github.com/tensorflow/tensorflow/issues?utf8=%E2%9C%93&q=is%3Aissue%20%22No%20Android%20SDK%20found%22%20is%3Aclosed%20
) (I tried various Android build searches too and they seemed to be different problems):

### Environment info
Operating System: Mac OS X 10.10.5

Installed version of CUDA and cuDNN: None

If installed from source, provide 

1. The commit hash (`git rev-parse HEAD`)
```
19d932f4cb49182c111d646267ac6c0bbc4f2f00
```
I cloned (with submodules) today, so this is the current master code.

2. The output of `bazel version`
```
Build label: 0.4.4-homebrew
Build target: bazel-out/local-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Thu Feb 2 01:06:38 2017 (1485997598)
Build timestamp: 1485997598
Build timestamp as int: 1485997598
```

### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code)

Here's the error:

> Daniels-MacBook-Air:tensorflow Dan$ bazel build //tensorflow/examples/android:tensorflow_demo
> ...
> <snip>
> ...
> ERROR: /private/var/tmp/_bazel_Dan/e9cc2fe61f735d74f6a33af534dffc47/external/bazel_tools/src/tools/android/java/com/google/devtools/build/android/incrementaldeployment/BUILD:3:1: in android_library rule @bazel_tools//src/tools/android/java/com/google/devtools/build/android/incrementaldeployment:incremental_stub_application: **No Android SDK found. Use the --android_sdk command line option to specify one.**

Here's the relevant section of my WORKSPACE file (Bazel 0.4.4 seems to require Android build tools 24.0.3 or newer, which is why I updated that too):
```
android_sdk_repository(
    name = "androidsdk",
    api_level = 23,
    build_tools_version = "24.0.3",
    # Replace with path to Android SDK on your system
    path = "/Developer/Android/sdk/",
)

android_ndk_repository(
    name="androidndk",
    path="/Developer/Android/android-ndk-r12b",
    api_level=21)
```
### What other attempted solutions have you tried?

I tried the path both with and without the trailing slash.  

I investigated how to pass the `--android_sdk` option to bazel based on [this Bazel documentation](https://bazel.build/versions/master/docs/command-line-reference.html):
```
--android_sdk=<a build target label> default: "@bazel_tools//tools/android:sdk"
Specifies Android SDK/platform that is used to build Android applications.
```
I couldn't figure out how to pass in that setting (plus I suspect it is supposed to come from the WORKSPACE file anyway).  Here's what I tried:
```
Daniels-MacBook-Air:tensorflow Dan$ bazel build --android_sdk=/Developer/Android/sdk //tensorflow/examples/android:tensorflow_demo
While parsing option --android_sdk=/Developer/Android/sdk: invalid label: /Developer/Android/sdk

Daniels-MacBook-Air:tensorflow Dan$ bazel build //tensorflow/examples/android:tensorflow_demo --android_sdk=/Developer/Android/sdk
While parsing option --android_sdk=/Developer/Android/sdk: invalid label: /Developer/Android/sdk
```

### Logs or other output that would be helpful
(If logs are large, please upload as attachment or provide link).

Full logs are here:
```
Daniels-MacBook-Air:tensorflow Dan$ bazel build //tensorflow/examples/android:tensorflow_demo
WARNING: Sandboxed execution is not supported on your system and thus hermeticity of actions cannot be guaranteed. See http://bazel.build/docs/bazel-user-manual.html#sandboxing for more information. You can turn off this warning via --ignore_unsupported_sandboxing.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:avgpooling_op.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:bounds_check.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:cwise_ops.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:cwise_ops_common.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:cwise_ops_gradients.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:eigen_activations.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:eigen_attention.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:eigen_backward_cuboid_convolutions.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:eigen_backward_spatial_convolutions.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:eigen_cuboid_convolution.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:eigen_patch_3d.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:eigen_pooling.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:eigen_softmax.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:eigen_spatial_convolutions.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:fifo_queue.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:maxpooling_op.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:ops_util.cc' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:ops_util.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:padding_fifo_queue.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:pooling_ops_common.cc' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:pooling_ops_common.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:queue_base.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:queue_op.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/kernels:typed_queue.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/util/ctc:ctc_beam_entry.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/util/ctc:ctc_beam_scorer.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/util/ctc:ctc_beam_search.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/util/ctc:ctc_decoder.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/util/ctc:ctc_loss_util.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/util/tensor_bundle:naming.cc' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/util/tensor_bundle:naming.h' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/util/tensor_bundle:tensor_bundle.cc' directly. You should either move the file to this package or depend on an appropriate rule there.
WARNING: /Users/Dan/Code/tensorflow/tensorflow/core/BUILD:826:12: in srcs attribute of cc_library rule //tensorflow/core:android_tensorflow_lib_lite: please do not import '//tensorflow/core/util/tensor_bundle:tensor_bundle.h' directly. You should either move the file to this package or depend on an appropriate rule there.
ERROR: /private/var/tmp/_bazel_Dan/e9cc2fe61f735d74f6a33af534dffc47/external/bazel_tools/src/tools/android/java/com/google/devtools/build/android/incrementaldeployment/BUILD:3:1: in android_library rule @bazel_tools//src/tools/android/java/com/google/devtools/build/android/incrementaldeployment:incremental_stub_application: No Android SDK found. Use the --android_sdk command line option to specify one.
ERROR: Analysis of target '//tensorflow/examples/android:tensorflow_demo' failed; build aborted.
INFO: Elapsed time: 7.929s
```

**Comment by girving** [comment]
@petewarden Are you still the right person for Android issues?

**Comment by andrewharp** [comment]
@daj When you run `/Developer/Android/sdk/tools/android` do you see the specified build tools and  API installed?

**Comment by daj** [comment]
I think I have everything I need for API level 21, API level 23 and build tools 24.0.3:
```
$ /Developer/Android/sdk/tools/android
```
![screen shot 2017-02-06 at 8 55 46 pm](https://cloud.githubusercontent.com/assets/739125/22674612/e5900754-ecae-11e6-80a9-c45e47bc2a6c.png)

![screen shot 2017-02-06 at 9 01 36 pm](https://cloud.githubusercontent.com/assets/739125/22674719/81d76d14-ecaf-11e6-905e-11d6cc45bc02.png)

![screen shot 2017-02-06 at 8 56 08 pm](https://cloud.githubusercontent.com/assets/739125/22674616/e91d51ce-ecae-11e6-982e-a0af6313c83d.png)

![screen shot 2017-02-06 at 8 56 20 pm](https://cloud.githubusercontent.com/assets/739125/22674619/eb3909f8-ecae-11e6-902d-cacf2276c4e7.png)

Sorry if I am missing something obvious...  I did a `bazel clean` and re-ran `bazel build //tensorflow/examples/android:tensorflow_demo` after each extra item I installed.

**Comment by daj** [comment]
D'oh, it looks like I was missing the specific API level that I had specified in the WORKSPACE file.  My bad.
![screen shot 2017-02-06 at 8 44 56 pm](https://cloud.githubusercontent.com/assets/739125/22674827/279d0e66-ecb0-11e6-97a1-5df9de45140b.png)

In hindsight the error message is pretty good, though it would be awesome if it said which specific SDK version was missing, e.g. `Android SDK version 23 not found`.  In my case I had almost all the other SDK versions installed, just not the one I had configured in the WORKSPACE file.

I did have the right build tools version installed (showing screenshot here in case it helps future people who find this GitHub issue):
![screen shot 2017-02-06 at 8 55 46 pm](https://cloud.githubusercontent.com/assets/739125/22674842/46c7209c-ecb0-11e6-8749-62cb33351942.png)

If there's a way I can improve the error message, let me know and I'll PR it (though if that requires a Bazel change then that might be beyond me!).
